skate-plugin/src/main/kotlin/com/slack/sgp/intellij/tracing/SkateTraceReporter.kt

```diff
@@ -0,0 +1,113 @@
+import com.intellij.openapi.application.ApplicationInfo
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.extensions.PluginId
+import com.intellij.openapi.project.Project
+import com.slack.sgp.intellij.util.tracingEndpoint
+import com.slack.sgp.tracing.KeyValue
+import com.slack.sgp.tracing.ListOfSpans
+import com.slack.sgp.tracing.model.buildSpan
+import com.slack.sgp.tracing.model.tagBuilderImpl
+import com.slack.sgp.tracing.reporter.SimpleTraceReporter
+import com.slack.sgp.tracing.reporter.TraceReporter
+import com.slack.sgp.tracing.reporter.TraceReporter.NoOpTraceReporter
+import java.time.Duration
+import java.time.Instant
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+import kotlinx.coroutines.runBlocking
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import org.apache.http.HttpException
+
+class SkateTraceReporter(private val project: Project, private val offline: Boolean = false) :
+  TraceReporter {
+
+  private val delegate by lazy {
+    val okHttpClient = lazy { OkHttpClient.Builder().build() }
+    val loggingInterceptor =
+      HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BODY }
+
+    val loggingClient = lazy {
+      okHttpClient.value.newBuilder().addInterceptor(loggingInterceptor).build()
+    }
+    if (offline) {
+      NoOpTraceReporter
+    } else {
+      val endPoint =
+        project.tracingEndpoint()
+          ?: throw RuntimeException("Required tracing endpoint to upload analytics")
+      SimpleTraceReporter(endPoint, loggingClient)
+    }
+  }
+
+  override suspend fun sendTrace(spans: ListOfSpans) {
+    try {
+      delegate.sendTrace(spans)
+      LOG.info(spans.toString())
+    } catch (httpException: HttpException) {
+      LOG.error("Uploading Skate plugin trace failed.", httpException)
+    }
+  }
+
+  fun createPluginUsageTraceAndSendTrace(
+    spanName: String,
+    startTimestamp: Instant,
+    spanDataMap: List<KeyValue>,
+    ideVersion: String = ApplicationInfo.getInstance().fullVersion,
+    skatePluginVersion: String? =
+      PluginManagerCore.getPlugin(PluginId.getId("com.slack.intellij.skate"))?.version
+  ): ListOfSpans? {
+    if (spanDataMap.isEmpty() || skatePluginVersion.isNullOrBlank()) {
+      return null
+    }
+    val traceTags =
+      tagBuilderImpl().apply {
+        "service_name" tagTo SERVICE_NAME
+        "database" tagTo DATABASE_NAME
+      }
+    val span =
+      buildSpan(
+        name = spanName,
+        startTimestampMicros =
+          startTimestamp.toEpochMilli().toDuration(DurationUnit.MILLISECONDS).inWholeMicroseconds,
+        durationMicros =
+          Duration.between(startTimestamp, Instant.now())
+            .toMillis()
+            .toDuration(DurationUnit.MILLISECONDS)
+            .inWholeMicroseconds
+      ) {
+        "skate_version" tagTo skatePluginVersion
+        "ide_version" tagTo ideVersion
+        "user" tagTo System.getenv("USER")
```

Comment:
This can get overridden to be an empty string, so let's check if it's blank and omit it if so

```suggestion
        System.getenv("USER").takeUnless { it.isBlank() }?.let { "user" tagTo it }
```